src/core/lib/slice/slice_buffer.cc

```diff
@@ -262,54 +276,85 @@ void grpc_slice_buffer_move_into(grpc_slice_buffer* src,
+          &src->slices[0], GRPC_SLICE_LENGTH(src->slices[0]) - bytes_extra);
+    }
+  } else {
+    // The whole thing.
+    dst->slices[dst->count] = src->slices[0];
+    ++src->slices;
+    --src->count;
+  }
+  ++dst->count;
+}
+
+template <bool incref>
 static void slice_buffer_move_first_maybe_ref(grpc_slice_buffer* src, size_t n,
-                                              grpc_slice_buffer* dst,
-                                              bool incref) {
-  GPR_ASSERT(src->length >= n);
-  if (src->length == n) {
+                                              grpc_slice_buffer* dst) {
+  if (n == 0) {
+    return;
+  }
+  if (n == src->length) {
     grpc_slice_buffer_move_into(src, dst);
     return;
   }
+  GPR_ASSERT(src->length > n);
+
+  size_t slices_to_copy = 0;
+  size_t bytes_so_far = 0;
+  size_t curr_idx = 0;
+  /* There are two general ways to architect this kind of method; we can copy a
+   * slice at a time, or we can pre-iterate the slices to see how many slices in
+   * total we need to copy. Given that the last slice may need to be split, the
+   * former approach suffers from the need for an additional comparison per
+   * slice to see if such a split is necessary. Pre-iterating as we've done here
+   * does require iterating the array twice, but saves the branches. */
+  while (bytes_so_far < n) {
+    bytes_so_far += GRPC_SLICE_LENGTH(src->slices[curr_idx]);
+    slices_to_copy++;
+    curr_idx++;
+  }
+  const size_t bytes_extra = bytes_so_far - n;
+  maybe_embiggen<true>(dst, slices_to_copy);
 
-  size_t output_len = dst->length + n;
-  size_t new_input_len = src->length - n;
-
-  while (src->count > 0) {
-    grpc_slice slice = grpc_slice_buffer_take_first(src);
-    size_t slice_len = GRPC_SLICE_LENGTH(slice);
-    if (n > slice_len) {
-      grpc_slice_buffer_add(dst, slice);
-      n -= slice_len;
-    } else if (n == slice_len) {
-      grpc_slice_buffer_add(dst, slice);
-      break;
-    } else if (incref) { /* n < slice_len */
-      grpc_slice_buffer_undo_take_first(
-          src, grpc_slice_split_tail_maybe_ref(&slice, n, GRPC_SLICE_REF_BOTH));
-      GPR_ASSERT(GRPC_SLICE_LENGTH(slice) == n);
-      grpc_slice_buffer_add(dst, slice);
-      break;
-    } else { /* n < slice_len */
-      grpc_slice_buffer_undo_take_first(
-          src, grpc_slice_split_tail_maybe_ref(&slice, n, GRPC_SLICE_REF_TAIL));
-      GPR_ASSERT(GRPC_SLICE_LENGTH(slice) == n);
-      grpc_slice_buffer_add_indexed(dst, slice);
-      break;
-    }
+  if (slices_to_copy == 1) {
+    slice_buffer_move_single_slice<incref>(src, dst, bytes_extra);
+  } else {
+    // memcpy all but the last slice to dst.
+    const size_t all_but_last = slices_to_copy - 1;
+    memcpy(&dst->slices[dst->count], &src->slices[0],
+           sizeof(grpc_slice) * all_but_last);
+    dst->count += all_but_last;
+    src->count -= all_but_last;
+    src->slices += all_but_last;
+    slice_buffer_move_single_slice<incref>(src, dst, bytes_extra);
```

Comment:
do we need this if `bytes_so_far == n`? we can probably just copy right?

Reply:
Should be able to. Theoretically grpc_slice_buffer_move_into can make things faster, but I haven't checked to see if it's worth it.